### PR TITLE
Add the ability to parse user attribute types

### PR DESCRIPTION
**Pull request by bowiec**
This adds the ability to parse user attributes into types based on the new `parseUserAttributes` setting. By default, it will maintain the existing behavior and forward user attributes as strings.

### CODE REVIEW

File: iterable-java-sdk/src/test/java/com/mparticle/iterable/IterableServiceTest.java [hunk 1]
```diff
@@ -94,7 +94,7 @@ public void testUserUpdate() throws Exception {
         UserUpdateRequest userUpdateRequest = new UserUpdateRequest();
         userUpdateRequest.userId = TEST_USER_ID;
         userUpdateRequest.email = TEST_EMAIL;
-        Map<String, String> attributes = new HashMap<String, String>();
+        Map<String, Object> attributes = new HashMap<String, Object>();
         attributes.put("test attribute key", "test attribute value");
```
Comment (samdozor) [inline iterable-java-sdk/src/test/java/com/mparticle/iterable/IterableServiceTest.java]:
```suggestion
        attributes.put("test string key", "test attribute string value");
        attributes.put("test boolean key", true);
        attributes.put("test int key", 1234);
        attributes.put("test double key", 1234.56D);
```

File: iterable-extension/src/main/java/com/mparticle/ext/iterable/IterableExtension.java [hunk 2]
```diff
@@ -244,6 +245,26 @@ void updateUser(EventProcessingRequest request) throws IOException {
         }
     }
 
+    private Map<String, Object> getUserAttributes(EventProcessingRequest request) {
+        String settingValue = request.getAccount().getAccountSettings().get(SETTING_PARSE_USER_ATTRIBUTES);
+        Boolean parseUserAttributes = Boolean.parseBoolean(settingValue);
```
Comment (samdozor) [inline iterable-extension/src/main/java/com/mparticle/ext/iterable/IterableExtension.java]:
```suggestion
        boolean parseUserAttributes = Boolean.parseBoolean(settingValue);
```

nit ^^ no reason to "autobox" here, `parseBoolean` returns a primitive

File: iterable-extension/src/main/java/com/mparticle/ext/iterable/IterableExtension.java [hunk 3]
```diff
@@ -24,6 +24,7 @@
     public static final String SETTING_APNS_KEY = "apnsProdIntegrationName";
     public static final String SETTING_APNS_SANDBOX_KEY = "apnsSandboxIntegrationName";
     public static final String SETTING_LIST_ID = "listId";
+    public static final String SETTING_PARSE_USER_ATTRIBUTES = "parseUserAttributes";
```
Comment (samdozor) [inline iterable-extension/src/main/java/com/mparticle/ext/iterable/IterableExtension.java]: We may actually want to change this to something like "coerce strings to scalars" - I noticed that the same problem exists for attributes/`dataFields` of `CommerceItem`.

**Comment by vbabenkoru** [comment]
@bowiec Is this going to be a hidden setting, not user-accessible?

**Comment by bowiec** [comment]
@vbabenkoru 

> @bowiec Is this going to be a hidden setting, not user-accessible?

This is going to be a visible setting, unless you have concerns. We are also planning on enabling this by default for new connections.

**Comment by vbabenkoru** [comment]
@bowiec Isn't it required to add a new setting to `processRegistrationRequest` for it to appear? Just trying to understand how this works.

File: iterable-extension/src/main/java/com/mparticle/ext/iterable/IterableExtension.java [hunk 4]
```diff
@@ -244,6 +245,28 @@ void updateUser(EventProcessingRequest request) throws IOException {
...
 
+    private Map<String, Object> convertAttributes(Map<String, String> attributes, boolean coerceStringsToScalars) {
+        if (attributes == null) {
+            return null;
+        }
+        
+        if (coerceStringsToScalars) {
+            return attemptTypeConversion(attributes);
+        }
+        else {
```
Comment (samdozor) [inline iterable-extension/src/main/java/com/mparticle/ext/iterable/IterableExtension.java]:
```suggestion
        } else {
```

File: iterable-extension/src/main/java/com/mparticle/ext/iterable/IterableExtension.java [hunk 5]
```diff
@@ -244,6 +245,28 @@ void updateUser(EventProcessingRequest request) throws IOException {
...
+            mapObj.putAll(attributes);
+
+            return mapObj;
+        }
+    }
+
+    private static boolean shouldCoerceStrings(EventProcessingRequest request) {
+        String settingValue = request.getAccount().getAccountSettings().get(SETTING_COERCE_STRINGS_TO_SCALARS);
+        
+        return Boolean.parseBoolean(settingValue);
```
Comment (samdozor) [inline iterable-extension/src/main/java/com/mparticle/ext/iterable/IterableExtension.java]:
```suggestion
        if (settingValue == null) { 
            return false;
        }
        return Boolean.parseBoolean(settingValue);
```

**Comment by julianmclain** [comment]
Looks good to me. The only thing I would propose is applying this setting to the dataFields map in `TrackPurchaseRequest` too.

**Comment by julianmclain** [comment]
Actually, I just did a final sweep and there's a dataFields Map in `Device` and `IterableRequest` that could use the same treatment.

**Comment by julianmclain** [comment]
Thanks @bowiec

**Comment by vbabenkoru** [comment]
Do we also need to bump the version in `processRegistrationRequest`?

**Comment by bowiec** [comment]
> One question: since this introduces a new setting, and that setting does not exist in existing customer accounts, isn't it going to default to true for them as well?
> Or are you going to backfill all existing integrations with false before deploying this change?

Correct on both counts. We manually updated things on our end to handle this for existing customers.

> Do we also need to bump the version in processRegistrationRequest?

I'm not certain about that, but I don't think we will be re-running the module registration request for this change since we've updated things manually. That said, I'll update it quickly here to make sure there are no issues.

**Pull request merged**